wordpress-mobile/release-toolkit
# Make explicit the use of the parameter `GITHUB_TOKEN` when use `create_release`

# What does this solve?

The issue is about CI failing at the moment of doing a GitHub release, The reason for the failure was a 401 from the GitHub API. This happened because neither GHHELPER_ACCESS nor GITHUB_TOKEN was available in the environment at runtime, As stated in [Issue#416](https://github.com/wordpress-mobile/release-toolkit/issues/416).  

# How was it solved?

The solution proposed was to change the implementation to make the token mandatory at the `github_client level`, and make the `create_release` receive an optional argument that will serve as the de facto GitHubToken.  In addition, create a fallback mechanism to the `create_release` function to use the environment variable `GITHUB_TOKEN` when the parameter passed at the function is not present.

A more verbose API in the actions by adding parameters for the tokens they expect will make it more difficult to forget their use of them. Making them invisible to our users can cause errors like the one on the issue.

# Testing Instructions 

Given this is a helper change affecting the way we use the GitHub client, the only E2E way to do it is by triggering the Github release using either a Fastlane lane or Gradle function with the new parameter set. Also, those scenarios are covered by unit testing

The expected behaviors are:
- The release succeeds if you pass the token explicitly or have set it on the environment variables
- The release fails if neither the token is passed at the create_release nor is present at the env variables.

### Example of Lane with the new githubtoken param set

```
create_release(
    repository: GHHELPER_REPO,
    version: version,
    release_notes_file_path: File.join(PROJECT_ROOT_FOLDER, 'WordPress', 'Resources', 'release_notes.txt'),
    release_assets: archive_zip_path.to_s,
    prerelease: options[:beta_release],
    githubtoken: GITHUB_TOKEN
)
```
